### PR TITLE
Merge Sessions into Pytorch

### PR DESCRIPTION
**Pull request by derekchen14**
first draft at working neural_session that can chat with cmd_session bot

### CODE REVIEW

File: negotiation/evaluate.py [hunk 1]
```diff
@@ -48,8 +48,8 @@
         cuda.set_device(args.gpuid[0])
 
     # Load the model.
-    mappings, model, model_args = \
-        model_builder.load_test_model(args, dummy_args.__dict__)
+    mappings, model, model_args = model_builder.load_test_model(args.checkpoint_file,
+        use_gpu(args), dummy_args.__dict__)
```
Comment (hhexiy) [inline negotiation/evaluate.py]: Why do you need use_gpu here? It's handled by load_test_model.

File: negotiation/neural/evaluator.py [hunk 2]
```diff
@@ -2,7 +2,7 @@
 import os
 from itertools import count
 
-from onmt.Utils import use_gpu
+from cocoa.pt_model.util import use_gpu
```
Comment (hhexiy) [inline negotiation/neural/evaluator.py]: Is this use_gpu different from onmt use_gpu?

File: negotiation/neural/evaluator.py [hunk 3]
```diff
@@ -11,33 +11,27 @@
...
     group = parser.add_argument_group('Efficiency')
     group.add_argument('--batch-size', type=int, default=30,
                        help='Batch size')
     group.add_argument('--gpuid', default=[], nargs='+', type=int,
                        help="Use CUDA on the listed devices.")
 
-    group = parser.add_argument_group('Logging')
-    group.add_argument('--verbose', action="store_true",
-                       help='Print scores and predictions for each sentence')
-
```
Comment (hhexiy) [inline negotiation/neural/evaluator.py]: Why is this deleted?

File: negotiation/neural/generator.py [hunk 4]
```diff
@@ -125,6 +126,7 @@ def unbottle(m):
             # Get all the pending current beam words and arrange for forward.
             inp = var(torch.stack([b.get_current_state() for b in beam])
                       .t().contiguous().view(1, -1))
+            smart_inp = smart_variable(inp, dtype="var")
```
Comment (hhexiy) [inline negotiation/neural/generator.py]: why do you need the smart_inp here? this is handled by `var()` right?

Comment (hhexiy) [reply negotiation/neural/generator.py]: best not to change the onmt code

File: negotiation/neural/generator.py [hunk 5]
```diff
@@ -138,7 +140,7 @@ def unbottle(m):
 
             # Run one step.
             dec_out, dec_states, attn = self.model.decoder(
-                inp, memory_bank, dec_states, memory_lengths=memory_lengths)
+                smart_inp, memory_bank, dec_states, memory_lengths=memory_lengths)
```
Comment (hhexiy) [inline negotiation/neural/generator.py]: similarly, why smart_inp?

File: negotiation/neural/model_builder.py [hunk 6]
```diff
@@ -149,9 +149,8 @@ def make_decoder(opt, embeddings):
                              embeddings=embeddings)
 
 
-def load_test_model(opt, dummy_opt):
-    checkpoint = torch.load(opt.model,
-                            map_location=lambda storage, loc: storage)
+def load_test_model(model_path, uses_gpu, dummy_opt):
+    checkpoint = torch.load(model_path, map_location=lambda storage, loc: storage)
```
Comment (hhexiy) [inline negotiation/neural/model_builder.py]: why did you delete opt? there are some options that we want to specify at run time instead of reading from the model checkpoint

File: negotiation/neural/model_builder.py [hunk 7]
```diff
@@ -160,8 +159,7 @@ def load_test_model(opt, dummy_opt):
 
     mappings = read_pickle('{}/vocab.pkl'.format(model_opt.mappings))
 
-    model = make_base_model(model_opt, mappings,
-                            use_gpu(opt), checkpoint)
+    model = make_base_model(model_opt, mappings, uses_gpu, checkpoint)
```
Comment (hhexiy) [inline negotiation/neural/model_builder.py]: similarly, why remove opt

File: negotiation/sessions/neural_session.py [hunk 8]
```diff
@@ -190,3 +195,110 @@ def output_to_tokens(self, output_dict):
...
+                self.vocab, sort_by_length=False, cuda=self.cuda)
+
+    def _decoder_args(self, entity_tokens):
+        inputs = self._process_entity_tokens(entity_tokens, 'decoding')
+        decoder_args = {'inputs': inputs,
+                'last_inds': self._get_last_inds(inputs),
+                'init_state': self.decoder_state,
+                'textint_map': self.env.textint_map,
+                }
+        return decoder_args
```
Comment (hhexiy) [inline negotiation/sessions/neural_session.py]: is _decoder_args needed?

File: negotiation/sessions/neural_session.py [hunk 9]
```diff
@@ -190,3 +195,110 @@ def output_to_tokens(self, output_dict):
...
+        entity_tokens = self._build_target_tokens(data["predictions"][0][0])
+        return entity_tokens
+
+    def _pred_to_token(self, preds):
+        n_best_predictions = preds[0].pred_sents
+        entity_tokens = n_best_predictions[0] # select top response
+        entity_tokens, _ = EncDecEvaluator.pred_to_token(preds,
+                self.env.stop_symbol, self.env.remove_symbols,
+                self.env.textint_map)
+        return entity_tokens
```
Comment (hhexiy) [inline negotiation/sessions/neural_session.py]: is _pred_to_token needed?

File: negotiation/sessions/neural_session.py [hunk 10]
```diff
@@ -190,3 +195,110 @@ def output_to_tokens(self, output_dict):
...
+        for tok in pred:
+            raw_token = vocab.ind_to_word[tok]
+            # str() to convert Entity
+            string_token = str(raw_token.canonical) if is_entity(raw_token) else raw_token
+            tokens.append(string_token)
+
+            if tokens[-1] == markers.EOS:
+                tokens = tokens[:-1]
+                break
+        return tokens
```
Comment (hhexiy) [inline negotiation/sessions/neural_session.py]: can you call UtteranceBuilder here instead of duplicating the function

File: negotiation/neural/evaluator.py (hunk 3 above)
Comment (derekchen14) [reply negotiation/neural/evaluator.py]: Adding this arg conflicts with the "verbose" option already included in `generate_dataset.py`.  I added this back in and resolved by using "conflict_handler" in argparse

**Comment by derekchen14** [comment]
The args being passed into `load_test_model` were changed because I wanted to minimize the options being passed around since there were default args, model args, checkpoint args, and dummy args.  It's not a big deal either way, I've change the API back to its original form.

Generally speaking, I tried to transition away from OMNT and instead incorporate what we need into cocoa.  Based on your comments though, I've reverted the OMNT-related portions back to its original form as much as possible.

Certain pytorch session methods were copied over from tensorflow session and no longer needed, so they have been removed.

File: scripts/generate_dataset.py [hunk 11]
```diff
@@ -16,7 +16,7 @@
 from core.controller import Controller
 from systems import add_system_arguments, get_system
 
-parser = argparse.ArgumentParser()
+parser = argparse.ArgumentParser(conflict_handler='resolve')
```
Comment (hhexiy) [inline scripts/generate_dataset.py:19]: is this because there are multiple --verbose?

Comment (derekchen14) [reply scripts/generate_dataset.py:19]: Yes, that is exactly right.  There is a `verbose` in generate_dataset and one in neural/evalutor.  Removing the `verbose` in generate_dataset causes other sessions to stop being verbose.  However, removing the `verbose` in neural/evaluator causes the "test" process to stop being verbose.  Thus, it seemed like conflict handler was the simplest way to handle this.
